README.md
# jmon-terraform-provider

Terraform provider for jmon

## Building and Installing

Install golang 1.15+

```
make
```

## Using in terraform

Define the provider:
```hcl
provider "jmon" {
  # This can be ommitted to default to http://localhost:5000
  url = "https://my-jmon-installation.com"
}

terraform {
  required_providers {
    jmon = {
      source  = "github.com/matthewjohn/jmon"
    }
  }
}
```

Create a minimal check:

```hcl
resource "jmon_check" "basic_check" {
  name = "My_Check"

  steps = <<EOF
# Check homepage
- goto: https://en.wikipedia.org/wiki/Main_Page
- check:
    title: Wikipedia, the free encyclopedia
EOF
}
```

An environment with a fully populated check
```hcl
resource "jmon_environment" "production" {
  name = "production"
}

resource "jmon_check" "full_check" {
  name = "Fully_Populated"

  environment         = jmon_environment.production.name
  timeout             = 60
  interval            = 300
  client              = "BROWSER_FIREFOX"
  screenshot_on_error = true

  steps = <<EOF
# Check homepage
- goto: https://en.wikipedia.org/wiki/Main_Page
- check:
    title: Wikipedia, the free encyclopedia
# Perform search
- find:
    id: searchform
    find:
    tag: input
    actions:
        - type: Pabalonium
        - press: enter
- check:
    url: "https://en.wikipedia.org/w/index.php?fulltext=Search&search=Pabalonium&title=Special%3ASearch&ns0=1"
- find:
    class: mw-search-nonefound
    check:
    text: There were no results matching the query.
- actions:
    - screenshot: Homepage
EOF
}

```


CHANGELOG.md
# Changelog

## [2.2.5](https://gitlab.dockstudios.co.uk/pub/jmon/jmon-terraform-provider/compare/v2.2.4...v2.2.5) (2023-08-26)


### Bug Fixes

* Fix removing attributes from a check not being detected and being set to default values. ([2d74852](https://gitlab.dockstudios.co.uk/pub/jmon/jmon-terraform-provider/commit/2d748524f640ff6c31328bb5d85394efa1119b20)), closes [#10](https://gitlab.dockstudios.co.uk/pub/jmon/jmon-terraform-provider/issues/10)
* Update Makefile to use registry.terraform.io rather than github to override default provider, rather than requiring a different provider path to be used ([a2685df](https://gitlab.dockstudios.co.uk/pub/jmon/jmon-terraform-provider/commit/a2685df0ec92123b05d13f1d2a6822884113d330)), closes [#10](https://gitlab.dockstudios.co.uk/pub/jmon/jmon-terraform-provider/issues/10)

## [2.2.3](https://gitlab.dockstudios.co.uk/pub/jmon/jmon-terraform-provider/compare/v2.2.2...v2.2.3) (2023-08-21)


### Bug Fixes

* Perform deep compare of steps structure before updating state to avoid non-functional changes showing differences in Terraform plan ([aa64397](https://gitlab.dockstudios.co.uk/pub/jmon/jmon-terraform-provider/commit/aa64397b501e178e6f121d6d4b7c2d3dcc04fc69)), closes [#7](https://gitlab.dockstudios.co.uk/pub/jmon/jmon-terraform-provider/issues/7)

## [2.2.2](https://gitlab.dockstudios.co.uk/pub/jmon/jmon-terraform-provider/compare/v2.2.1...v2.2.2) (2023-08-11)


### Bug Fixes

* **docs:** Fix example for JSON matching ([85f98dc](https://gitlab.dockstudios.co.uk/pub/jmon/jmon-terraform-provider/commit/85f98dc3f5bfdbb3948c972197fdb1ac91c78dd9))
* Fix call to provider ([d69e539](https://gitlab.dockstudios.co.uk/pub/jmon/jmon-terraform-provider/commit/d69e539db70c918c28f1c72b02305d64c8ff7876))

## [2.2.1](https://gitlab.dockstudios.co.uk/pub/jmon/jmon-terraform-provider/compare/v2.2.0...v2.2.1) (2023-08-11)


### Bug Fixes

* Fix references to matthewjohn, after moving repository to dockstudios ([3be1124](https://gitlab.dockstudios.co.uk/pub/jmon/jmon-terraform-provider/commit/3be11242c51dbc10f3106b6fbc21049a780e0a82)), closes [#5](https://gitlab.dockstudios.co.uk/pub/jmon/jmon-terraform-provider/issues/5)

# [2.2.0](https://gitlab.dockstudios.co.uk/pub/jmon/jmon-terraform-provider/compare/v2.1.1...v2.2.0) (2023-08-11)


### Features

* Add support for setting check attributes ([e6d5328](https://gitlab.dockstudios.co.uk/pub/jmon/jmon-terraform-provider/commit/e6d532806716e087628d84c264c34bef3d23bb9d)), closes [#6](https://gitlab.dockstudios.co.uk/pub/jmon/jmon-terraform-provider/issues/6)

# [2.1.0](https://gitlab.dockstudios.co.uk/pub/jmon/jmon-terraform-provider/compare/v2.0.0...v2.1.0) (2023-08-10)


### Bug Fixes

* Add headers to all requests made in check resource ([a30f4e4](https://gitlab.dockstudios.co.uk/pub/jmon/jmon-terraform-provider/commit/a30f4e479e32a2dbb1b936d95190a3e718399a78)), closes [#3](https://gitlab.dockstudios.co.uk/pub/jmon/jmon-terraform-provider/issues/3)
* Add semantic-release config ([c71fedb](https://gitlab.dockstudios.co.uk/pub/jmon/jmon-terraform-provider/commit/c71fedbc1991a4a7eb49d10d6f25c3a2c76c3a94)), closes [#5](https://gitlab.dockstudios.co.uk/pub/jmon/jmon-terraform-provider/issues/5)


### Features

* Add api_key attribute to provider and pass API key header in request headers ([dc882a5](https://gitlab.dockstudios.co.uk/pub/jmon/jmon-terraform-provider/commit/dc882a52e8e593ce5cc371ae431756032ae6dab6)), closes [#3](https://gitlab.dockstudios.co.uk/pub/jmon/jmon-terraform-provider/issues/3)


jmon/provider.go
package jmon

import (
	//"log"
	"context"
	"net/http"
	"time"

	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
)

func Provider() *schema.Provider {
	return &schema.Provider{
		ConfigureContextFunc: providerConfigure,

		Schema: map[string]*schema.Schema{
			"url": {
				Type:     schema.TypeString,
				Optional: true,
			},
			"api_key": {
				Type:     schema.TypeString,
				Optional: true,
			},
		},

		DataSourcesMap: map[string]*schema.Resource{},

		ResourcesMap: map[string]*schema.Resource{
			"jmon_environment": resourceEnvironment(),
			"jmon_check":       resourceCheck(),
		},
	}
}

type ProviderClient struct {
	url        string
	apiKey     string
	httpClient *http.Client
	headers    http.Header
}

func providerConfigure(ctx context.Context, d *schema.ResourceData) (interface{}, diag.Diagnostics) {
	var diags diag.Diagnostics
	client := &ProviderClient{}

	url := d.Get("url").(string)
	if url == "" {
		url = "http://localhost:5000"
	}
	client.url = url

	client.apiKey = d.Get("api_key").(string)

	client.headers = make(http.Header)
	client.headers.Set("Content-Type", "application/json")
	client.headers.Set("Accept", "application/json")
	if client.apiKey != "" {
		client.headers.Set("X-JMon-Api-Key", client.apiKey)
	}

	client.httpClient = &http.Client{
		Timeout: time.Second * 30,
	}

	return client, diags
}


jmon/resource_environment.go
package jmon

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io/ioutil"
	"log"
	"net/http"

	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
)

func resourceEnvironment() *schema.Resource {
	return &schema.Resource{
		CreateContext: resourceEnvironmentCreate,
		ReadContext:   resourceEnvironmentRead,
		// UpdateContext: resourceEnvironmentUpdate,
		DeleteContext: resourceEnvironmentDelete,

		Schema: map[string]*schema.Schema{
			"name": &schema.Schema{
				Type:     schema.TypeString,
				Required: true,
				ForceNew: true,
			},
		},
		Importer: &schema.ResourceImporter{
			StateContext: schema.ImportStatePassthroughContext,
		},
	}
}

func generateEnvironmentUrl(client *ProviderClient, name string) string {
	return fmt.Sprintf("%s/api/v1/environments/%s", client.url, name)
}

func getEnvironmentByName(d *schema.ResourceData, m interface{}, name string, responseBody *[]byte) (error, bool) {
	client := m.(*ProviderClient)

	// Create request to GET environment
	var url string = generateEnvironmentUrl(client, name)
	log.Printf("[jmon] Perform GET: %s", url)
	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		return err, false
	}

	// Add headers to request
	req.Header = client.headers.Clone()

	// Perform request
	r, err := client.httpClient.Do(req)
	if err != nil {
		return err, false
	}
	defer r.Body.Close()

	// Read response body
	*responseBody, err = ioutil.ReadAll(r.Body)
	if err != nil {
		return err, false
	}

	log.Printf("[jmon] Response code: %d", r.StatusCode)
	log.Printf("[jmon] Body: %s", responseBody)

	var exists bool
	exists = true
	// Check status code to determine if check exists
	if r.StatusCode == 404 {
		exists = false
	}

	return nil, exists
}

func resourceEnvironmentCreate(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
	var diags diag.Diagnostics

	// Determine if environment already exists
	var responseBody []byte
	err, exists := getEnvironmentByName(d, m, d.Get("name").(string), &responseBody)
	if err != nil {
		return diag.FromErr(err)
	}
	if exists {
		return diag.FromErr(errors.New(fmt.Sprintf("An environment already exists with the name: %s", d.Get("name").(string))))
	}

	// Create environment
	client := m.(*ProviderClient)
	var url string = fmt.Sprintf("%s/api/v1/environments", client.url)
	values := map[string]string{"name": d.Get("name").(string)}
	jsonPostData, err := json.Marshal(values)
	if err != nil {
		return diag.FromErr(err)
	}

	// Create HTTP request
	log.Printf("[jmon] Perform POST: %s: %s", url, jsonPostData)
	req, err := http.NewRequest("POST", url, bytes.NewBuffer(jsonPostData))
	if err != nil {
		return diag.FromErr(err)
	}

	// Add headers to request
	req.Header = client.headers.Clone()

	// Perform request
	r, err := client.httpClient.Do(req)
	if err != nil {
		return diag.FromErr(err)
	}
	defer r.Body.Close()

	// Read response body
	responseBody, err = ioutil.ReadAll(r.Body)
	if err != nil {
		return diag.FromErr(err)
	}

	log.Printf("[jmon] Response code: %d", r.StatusCode)
	log.Printf("[jmon] Body: %s", responseBody)

	if r.StatusCode != 200 {
		return diag.FromErr(errors.New(fmt.Sprintf("Got response code: %d from server: %s", r.StatusCode, string(responseBody))))
	}

	// Set ID of resource to environment name
	d.SetId(d.Get("name").(string))

	return diags
}

func resourceEnvironmentRead(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {

	var diags diag.Diagnostics

	var responseBody []byte
	err, exists := getEnvironmentByName(d, m, d.Get("name").(string), &responseBody)

	if err != nil {
		return diag.FromErr(err)
	}

	// If check does not exist, reset ID
	// and return early
	if exists == false {
		log.Printf("[jmon] Environment does not exist: %s", string(responseBody))
		d.SetId("")
		return diags
	}

	// Set name attribute from ID of object.
	// This is important during imports, as the name attribute
	// does not yet exist, so will be updated from the imported ID
	d.Set("name", d.Id())

	return diags
}

func resourceEnvironmentUpdate(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
	var diags diag.Diagnostics

	// There are no attributes that can be updated (yet),
	// so immediately return

	return diags
}

func resourceEnvironmentDelete(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
	var diags diag.Diagnostics

	client := m.(*ProviderClient)

	// Create DELETE request
	req, err := http.NewRequest("DELETE", generateEnvironmentUrl(client, d.Get("name").(string)), nil)
	if err != nil {
		return diag.FromErr(err)
	}

	// Add headers to request
	req.Header = client.headers.Clone()

	// Perform request
	r, err := client.httpClient.Do(req)
	if err != nil {
		return diag.FromErr(err)
	}
	defer r.Body.Close()

	// Read response body
	var responseBody []byte
	responseBody, err = ioutil.ReadAll(r.Body)
	if err != nil {
		return diag.FromErr(err)
	}

	// Check status code
	if r.StatusCode != 200 {
		return diag.FromErr(errors.New(fmt.Sprintf("[jmon] Environment failed to delete: %s", string(responseBody))))
	}

	return diags
}


jmon/resource_check.go
package jmon

import (
	"bytes"
	"context"
	"errors"
	"fmt"
	"io/ioutil"
	"log"
	"net/http"
	"reflect"
	"strings"

	"gopkg.in/yaml.v2"

	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
)

type CheckData struct {
	Name              string                 `yaml:"name"`
	Environment       string                 `yaml:"environment,omitempty"`
	Steps             []interface{}          `yaml:"steps"`
	ScreenshotOnError bool                   `yaml:"screenshot_on_error,omitempty"`
	Interval          int                    `yaml:"interval,omitempty"`
	Timeout           int                    `yaml:"timeout,omitempty"`
	Client            string                 `yaml:"client,omitempty"`
	Enable            bool                   `yaml:"enable,omitempty"`
	Attributes        map[string]interface{} `yaml:"attributes,omitempty"`
}

func resourceCheck() *schema.Resource {
	return &schema.Resource{
		CreateContext: resourceCheckCreate,
		ReadContext:   resourceCheckRead,
		UpdateContext: resourceCheckUpdate,
		DeleteContext: resourceCheckDelete,

		Schema: map[string]*schema.Schema{
			"name": &schema.Schema{
				Type:     schema.TypeString,
				Required: true,
				ForceNew: true,
			},
			"environment": &schema.Schema{
				Type:     schema.TypeString,
				Optional: true,
				ForceNew: true,
				Default:  "default",
			},
			"steps": &schema.Schema{
				Type:     schema.TypeString,
				Required: true,
			},
			"screenshot_on_error": &schema.Schema{
				Type:     schema.TypeBool,
				Optional: true,
				Computed: true,
			},
			"interval": &schema.Schema{
				Type:     schema.TypeInt,
				Optional: true,
				Default:  0,
			},
			"timeout": &schema.Schema{
				Type:     schema.TypeInt,
				Optional: true,
				Default:  0,
			},
			"client": &schema.Schema{
				Type:     schema.TypeString,
				Optional: true,
				Default:  nil,
			},
			"enable": &schema.Schema{
				Type:     schema.TypeBool,
				Optional: true,
				Default:  true,
			},
			"attributes": &schema.Schema{
				Type:     schema.TypeMap,
				Elem:     &schema.Schema{Type: schema.TypeString},
				Optional: true,
			},
		},
		Importer: &schema.ResourceImporter{
			StateContext: schema.ImportStatePassthroughContext,
		},
	}
}

func generateId(name string, environment string) string {
	var id string
	id = name
	if environment != "" {
		id = id + "/" + environment
	}
	return id
}

func nameEnvironmentFromId(id string) (string, string, error) {
	splitNames := strings.Split(id, "/")
	if len(splitNames) == 1 {
		return splitNames[0], "", nil
	} else if len(splitNames) == 2 {
		return splitNames[0], splitNames[1], nil
	} else {
		return "", "", errors.New("Cannot parse invalid ID")
	}
}

func generateCheckUrl(client *ProviderClient, name string, environment string) string {
	return fmt.Sprintf("%s/api/v1/checks/%s/environments/%s", client.url, name, environment)
}

func upsertCheck(d *schema.ResourceData, m interface{}, check *CheckData) error {

	// Convert resource attributes to attributes of check data object
	check.Name = d.Get("name").(string)
	check.Environment = d.Get("environment").(string)
	check.Interval = d.Get("interval").(int)
	check.Timeout = d.Get("timeout").(int)
	check.ScreenshotOnError = d.Get("screenshot_on_error").(bool)
	check.Client = d.Get("client").(string)
	check.Enable = d.Get("enable").(bool)
	check.Attributes = d.Get("attributes").(map[string]interface{})

	// Convert steps YAML to interface in check object
	ymlErr := yaml.Unmarshal([]byte(d.Get("steps").(string)), &check.Steps)
	if ymlErr != nil {
		return ymlErr
	}

	client := m.(*ProviderClient)

	// Convert entire check to YAML
	var yamlOutput []byte
	yamlOutput, err := yaml.Marshal(&check)
	if err != nil {
		return err
	}

	// Create reader for post data
	postDataReader := bytes.NewReader(yamlOutput)

	// Check request
	req, err := http.NewRequest("POST", fmt.Sprintf("%s/api/v1/checks", client.url), postDataReader)
	if err != nil {
		return err
	}

	// Add headers to request
	req.Header = client.headers.Clone()

	// Perform request
	r, err := client.httpClient.Do(req)
	if err != nil {
		return err
	}
	defer r.Body.Close()

	// Read response body
	var responseBody []byte
	responseBody, err = ioutil.ReadAll(r.Body)
	if err != nil {
		return err
	}

	// Check status code
	if r.StatusCode != 200 {
		return errors.New(fmt.Sprintf("Check failed to create/update: %s", string(responseBody)))
	}

	return nil
}

func resourceCheckCreate(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
	var diags diag.Diagnostics

	// Determine if check already exists
	var responseBody []byte
	err, exists := getCheckByNameAndEnvironment(d, m, d.Get("name").(string), d.Get("environment").(string), &responseBody)
	if err != nil {
		return diag.FromErr(err)
	}
	if exists {
		return diag.FromErr(errors.New(fmt.Sprintf("A check already exists with the name: %s", d.Get("name").(string))))
	}

	var check CheckData
	err = upsertCheck(d, m, &check)

	if err != nil {
		return diag.FromErr(err)
	}

	// Set ID of resource to check name
	d.SetId(generateId(check.Name, check.Environment))

	return diags
}

func getCheckByNameAndEnvironment(d *schema.ResourceData, m interface{}, name string, environment string, responseBody *[]byte) (error, bool) {
	client := m.(*ProviderClient)

	// Check request
	var url string = generateCheckUrl(client, name, environment)
	log.Printf("[jmon] Perform GET: %s", url)
	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		return err, false
	}

	// Add headers to request
	req.Header = client.headers.Clone()

	// Perform request
	r, err := client.httpClient.Do(req)
	if err != nil {
		return err, false
	}
	defer r.Body.Close()

	// Read response body
	*responseBody, err = ioutil.ReadAll(r.Body)
	if err != nil {
		return err, false
	}

	log.Printf("[jmon] Response code: %d", r.StatusCode)
	log.Printf("[jmon] Body: %s", responseBody)

	var exists bool
	exists = true
	// Check status code to determine if check exists
	if r.StatusCode == 404 {
		exists = false
	}

	return nil, exists
}

func resourceCheckRead(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {

	var diags diag.Diagnostics

	name, environment, err := nameEnvironmentFromId(d.Id())
	if err != nil {
		return diag.FromErr(err)
	}
	// Set default environment, if it does not exist
	if environment == "" {
		environment = "default"
		d.SetId(generateId(name, "default"))
	}

	var responseBody []byte
	err, exists := getCheckByNameAndEnvironment(d, m, name, environment, &responseBody)

	if err != nil {
		return diag.FromErr(err)
	}

	// If check does not exist, reset ID
	// and return early
	if exists == false {
		log.Printf("[jmon] Check does not exist: %s", string(responseBody))
		d.SetId("")
		return diags
	}

	var check CheckData
	err = yaml.Unmarshal(responseBody, &check)
	log.Printf("[jmon] Unmarshalled to: %v", check)
	if err != nil {
		return diag.FromErr(err)
	}

	// Only set steps string from API response, if the strctures differ.
	// If they aren't different, do not update, comments and insignificant
	// whitespace changes will cause a different in value, even if the
	// generated structure is fundermentally the same.
	updateSteps := true

	// Unmarshall current steps and perform deep compare
	var currentSteps []interface{}
	ymlErr := yaml.Unmarshal([]byte(d.Get("steps").(string)), &currentSteps)
	if ymlErr == nil {
		if reflect.DeepEqual(currentSteps, check.Steps) {
			updateSteps = false
		}
	}

	if updateSteps {
		stepsString, err := yaml.Marshal(&check.Steps)
		if err != nil {
			return diag.FromErr(err)
		}
		d.Set("steps", string(stepsString))
	}

	// Set name attribute from ID of object.
	// This is important during imports, as the name attribute
	// does not yet exist, so will be updated from the imported ID
	d.Set("name", name)
	d.Set("environment", environment)

	d.Set("interval", check.Interval)
	d.Set("timeout", check.Timeout)
	d.Set("client", check.Client)
	d.Set("screenshot_on_error", check.ScreenshotOnError)
	d.Set("enable", check.Enable)
	d.Set("attributes", check.Attributes)

	return diags
}

func resourceCheckUpdate(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
	var diags diag.Diagnostics

	var check CheckData
	err := upsertCheck(d, m, &check)

	if err != nil {
		return diag.FromErr(err)
	}

	return diags
}

func resourceCheckDelete(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
	var diags diag.Diagnostics

	client := m.(*ProviderClient)

	// Check request
	req, err := http.NewRequest("DELETE", generateCheckUrl(client, d.Get("name").(string), d.Get("environment").(string)), nil)
	if err != nil {
		return diag.FromErr(err)
	}

	// Add headers to request
	req.Header = client.headers.Clone()

	// Perform request
	r, err := client.httpClient.Do(req)
	if err != nil {
		return diag.FromErr(err)
	}
	defer r.Body.Close()

	// Read response body
	var responseBody []byte
	responseBody, err = ioutil.ReadAll(r.Body)
	if err != nil {
		return diag.FromErr(err)
	}

	// Check status code
	if r.StatusCode != 200 {
		return diag.FromErr(errors.New(fmt.Sprintf("[jmon] Check failed to delete: %s", string(responseBody))))
	}

	return diags
}
